File: app/services/nvidia_service.py
import logging
import random
try:
    import pynvml
    HAS_NVML = True
except ImportError:
    HAS_NVML = False

logger = logging.getLogger("app.nvidia")

class NvidiaService:
    def __init__(self):
        self.initialized = False
        if HAS_NVML:
            try:
                pynvml.nvmlInit()
                self.initialized = True
                logger.info("NVML Initialized successfully.")
            except Exception as e:
                logger.warning(f"NVML Init failed (likely no driver): {e}")
                self.initialized = False

    def get_gpu_count(self):
        if not self.initialized:
            return 1 # Simulate 1 GPU if not found
        try:
            return pynvml.nvmlDeviceGetCount()
        except:
            return 0

    def get_gpu_info(self, index):
        """Returns static info about a GPU."""
        if not self.initialized:
            return {
                "name": "NVIDIA GeForce RTX 4090 (Simulated)",
                "uuid": f"GPU-SIM-{index}",
                "vram_total_mb": 24576
            }
        
        try:
            handle = pynvml.nvmlDeviceGetHandleByIndex(index)
            name = pynvml.nvmlDeviceGetName(handle)
            uuid = pynvml.nvmlDeviceGetUUID(handle)
            mem = pynvml.nvmlDeviceGetMemoryInfo(handle)
            return {
                "name": name,
                "uuid": uuid,
                "vram_total_mb": mem.total // (1024 * 1024)
            }
        except Exception as e:
            logger.error(f"Error getting GPU info for index {index}: {e}")
            return None

    def get_gpu_stats(self, index):
        """Returns dynamic stats for a GPU."""
        if not self.initialized:
            # Random simulation for NMS demo
            return {
                "utilization_gpu": random.randint(10, 80),
                "utilization_mem": random.randint(20, 60),
                "vram_used_mb": random.randint(2000, 12000),
                "temperature": random.randint(45, 75),
                "power_draw_w": random.uniform(50.0, 350.0)
            }
        
        try:
            handle = pynvml.nvmlDeviceGetHandleByIndex(index)
            util = pynvml.nvmlDeviceGetUtilizationRates(handle)
            mem = pynvml.nvmlDeviceGetMemoryInfo(handle)
            temp = pynvml.nvmlDeviceGetTemperature(handle, pynvml.NVML_TEMPERATURE_GPU)
            try:
                power = pynvml.nvmlDeviceGetPowerUsage(handle) / 1000.0
            except:
                power = 0.0
            
            return {
                "utilization_gpu": util.gpu,
                "utilization_mem": util.memory,
                "vram_used_mb": mem.used // (1024 * 1024),
                "temperature": temp,
                "power_draw_w": power
            }
        except Exception as e:
            logger.error(f"Error getting GPU stats for index {index}: {e}")
            return None

nvidia_service = NvidiaService()

File: crm-portal-ts/src/services/nvidia_service.ts
import { createLogger } from '../utils'; // Assuming you have an utils module to handle logging in NodeJS environment, replace this import accordingly if not available.
const logger = createLogger('app', true);  
interface GpuInfo{
    name: string; 
    uuid: string;    
    vram_total_mb : number ;      // Assuming all GPUs have the same amount of VRAM, replace this with actual data if not.      
}
class NvidiaService {  
    private initialized = false;       
    constructor()  {         
         try{            
            pynvml?.nvmlInit();          
              logger.info("NVML Initialized successfully.");     }catch(e){logger.warning('Failed to initialize NVML: ' + e); this.initialized=false;}   },      // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    gpuCount = (): number => {             return !this.initialized ? 1 : pynvml?.nvmlDeviceGetCount() || 0;},     },      // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getGpuInfo = async (index: number): Promise<GpuInfo | null> => {             try{              const handle  = pynvml?.nvmlDeviceGetHandleByIndex(index);               return !handle ?  null :   await this.getGpuDetailsFromNvidiaAPI(handle) ;}catch(e){logger.error('Error getting GPU info for index ' + e );return null;} },      // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getGpuInfoDetailsFromNvidiaAPI = async (handle: any): Promise<any> => {             const name  = pynvml?.nvmlDeviceGetName(handle);               return  await this.getMemoryAndUtilizationRatesForHandle(handle) ;},     // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getGpuDetailsFromNvidiaAPI = async (handle: any): Promise<any> => {             const mem  = await this.getMemoryAndUtilizationRatesForHandle(handle);               return  {"name": name, "uuid" : pynvml?.nvmlDeviceGetUUID(handle), ...mem};},     // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getMemoryAndUtilizationRatesForHandle = async (handle: any): Promise<any> => {             const util  = await pynvml?.nvmlDeviceGetUtilizationRates(handle);               return  {"utilization_gpu" : util.gpu, "utilization_mem": util.memory , ...await this.getTemperatureAndPowerUsageForHandle(handle)};},     // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getTemperatureAndPowerUsageForHandle = async (handle: any): Promise<any> => {             const temp  = pynvml?.nvmlDeviceGetTemperature(handle, 1024);               return   {"temperature" :temp , ...await this.getPowerDrawFromNvidiaAPI(handle)};},     // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
    getPowerDrawFromNvidiaAPI = async (handle: any): Promise<any> => {             const power  = pynvml?.nvmlDeviceGetPowerUsage(handle) / 1000.0;               return   {"power_draw" : power };},     // Assuming we have a global variable named "pynvml", replace with actual data if not available in NodeJS environment
}    }